index.rst

```diff
@@ -1,57 +1,236 @@
+This event will also include a unique identifier for the visit.
+This unique identifier may be arbitrary; it need not be related to any particular exposure identifier or observing script identifier, but it must be made available to the command that is transferring the image from the Camera systems to Data Management.
 
-     Caption text.
+Latencies and overheads must be minimized.
+Since the Alert Production timeline is only 60 seconds from close of shutter on the last exposure of the visit, latencies on the order of seconds and potentially tenths of seconds do matter.
+As much I/O and computation as possible should be done after ``next_visit`` arrives before the exposure images do.
+Ingest of image metadata and starting pipelines should be as efficient as possible.
 
-   Run: ``make html`` and ``open _build/html/index.html`` to preview your work.
-   See the README at https://github.com/lsst-sqre/lsst-technote-bootstrap or
-   this repo's README for more info.
+The framework should execute in parallel, with a minimum of one process per CCD/detector.
+Amplifier-level or Source-level parallelism may be desirable for some functions.
 
-   Feel free to delete this instructional comment.
+It should handle single and multi-exposure visits, with the number of snaps per visit ideally known only at ``next_visit`` time.
 
-:tocdepth: 1
+It must be able to execute different pipelines on different images, at a minimum via configuration but ideally on a per-visit basis at ``next_visit`` time.
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+Since pipeline execution takes longer than the interval between visits, the framework must be able to have more than one visit in flight at a time.
+As some visits may take longer than others due to crowding, artifacts, or other influences, it must be elastic to allow compute resources to be used as needed.
 
-.. sectnum::
+It must be resilient to failures of communication, aborted visits, loss of exposures, memory exhaustion, and other problems.
+But occasional delays in executing or even outright failures to execute the pipeline for a visit are acceptable, as daytime catch-up processing is planned.
 
-.. TODO: Delete the note below before merging new content to the main branch.
 
-.. note::
+Initial Design
+==============
 
-   **This technote is not yet published.**
+The DDS ``next_visit`` event is translated to an HTTP POST containing the same information.
+One POST message is issued per CCD.
+That message is used to reserve a worker to process the entire visit.
+The HTTP message handler does not return until the visit has been successfully processed or a failure has occurred, been identified, and reported.
+
+Within the message handler, a separate messaging system is configured to receive notifications of exposure image arrival for the visit.
```

Comment:
This design seems to defeat most of the advantages of a message/event-based architecture, especially given the later remark that "listening for HTTP events in an HTTP event handler is tricky."

Why can't the exposure event handler be outside the `next_visit` event handler, and respond by sending a message (which the pipeline worker has been configured to listen for)? This gives a natural, concurrent way to handle control flow instead of having the `next_visit` handler need a homebrew signaling system. I think it also makes it easier to handle missed messages in either `next_visit` or `exposure_arrived`.

Reply:
Part of this is so that state doesn't need to persist outside the main `next_visit` event handler.  If the `next_visit` handler returned after preparing the local Butler and a different handler was used for `exposure_arrived`, we'd need to persist state between the handlers as well as ensure that `exposure_arrived` ends up at the same place as the `next_visit`.

I'm not exactly sure what you're getting at by having an "outside" event handler that sends a message which needs to be listened for.  That's exactly what is happening here; the "worker thread" in the `next_visit` handler prepares the local Butler and then listens for the exposures to arrive.

Missed `next_visit` events are basically unrecoverable by Prompt Processing; they have to be picked up by catch-up processing.  Missed `exposure_arrived` events could be mitigated by checking for exposure existence just before timing out.  Of course if the exposures themselves don't arrive (but were taken), that would also have to be picked up by catch-up processing.